### PR TITLE
Issue-14: create adapter framework for email providers

### PR DESCRIPTION
**Pull request by mogmarsh**
### Related Issue
Issue: [#14](https://github.com/alleyinteractive/wp-newsletter-builder/issues/14)

### Description
This Pull Request implements the adapter framework for different email providers, making it easy to switch between different email providers such as CampaignMonitor, Omeda, SendGrid, etc.

An example of this type of architecture can be found in [Elastic Search Extensions](https://github.com/alleyinteractive/elasticsearch-extensions/tree/main/lib/adapters)

### Use Case
Developers should be able to specify the provider to use via a filter.

### Changes
- [x] Implement adapter framework for email providers
- [x] Add filter to allow developers to specify email provider

### Testing
- [x] Test with different email providers
- [x] Ensure backward compatibility with existing code
- [x] Add unit tests for new framework
- [x] Update documentation for developers on how to use the new adapter framework and filter

### CODE REVIEW

File: plugins/newsletter-from-post/index.php [hunk 1]
```diff
@@ -38,12 +38,29 @@ function action_enqueue_post_sidebar_assets() {
...
+		$template_map[ $template->ID ] = $template->post_title;
+	}
+
 	wp_enqueue_script( 'plugin-newsletter-from-post' );
 	wp_localize_script(
 		'plugin-newsletter-from-post',
 		'newsletterBuilder',
 		[
-			'fromNames'     => Campaign_Monitor_Client::instance()->get_from_names(),
+			'fromNames'     => $settings->get_from_names(),
```
Comment (attackant) [inline plugins/newsletter-from-post/index.php:62]: Would a more descriptive variable name for $settings be helpful? Maybe $email_provider or $provider_settings ?

File: plugin.php [hunk 2]
```diff
@@ -58,18 +58,53 @@ function () {
...
+					</p>
+				</div>
+				<?php
+			}
+		);
+
+		return;
+	}
+	global $newsletter_builder_email_provider;
+	$provider                          = new $selected_email_provider();
```
Comment (attackant) [inline plugin.php]: To my taste adding a blank line below this so the formatter doesn't add so many spaces is more readable.

File: src/email-providers/interface-email-provider.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	 * @param array  $custom_fields The custom fields.
+	 * @return array|false
+	 */
+	public function add_subscriber( $list_id, $email, $custom_fields = [] );
+	/**
+	 * Remove subscriber from list.
+	 *
+	 * @param string $list_id The list id.
+	 * @param string $email The email address.
+	 * @return array|false
```
Comment (attackant) [inline src/email-providers/interface-email-provider.php:96]: Just wondering when we might choose false versus a WP_error?

**Review by attackant — APPROVED** [review]
Looks good to me, made a few suggestions and asked a question 🐙

File: src/class-settings.php [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	 * Gets the lists for the client.
+	 *
+	 * @TODO: Add caching that works on Pantheon and WordPress VIP.
+	 *
+	 * @return array|false
+	 */
+	public function get_lists() {
+		$settings = get_option( static::SETTINGS_KEY );
+		if ( empty( $settings ) || empty( $settings['api_key'] ) || empty( $settings['client_id'] ) ) {
+			return false;
```
Comment (attackant) [inline src/class-settings.php:135]: There is some duplication fetching the settings that could be extracted into a private method potentially throughout this file.

File: plugins/newsletter-from-post/index.php (hunk 1 above)
Comment (mogmarsh) [reply plugins/newsletter-from-post/index.php:62]: @attackant I moved the general settings to be separate from the email provider, so it is just settings and it's all in class Settings.

**Pull request merged**